idl2matlab.1
.TH "idl2matlab" "1" "1.5 (21 January 2009)" "ILL (richard@ill.fr and farhi@ill.fr)" "USER COMMANDS"
.SH "NAME"
idl2matlab \- This tool translates automatically an IDL code into Matlab/Scilab code.
.SH "SYNOPSIS"
.B idl2matlab
[\-options] inputFile [outputDirectory]
.SH "DESCRIPTION"
The
.B idl2matlab
command is an automatic translation tool from the IDL (and partly PV\-Wave) to Matlab and Scilab source code.
.PP 
idl(1) [http://www.ittvis.com] and matlab(1) [http://www.mathworks.com] are two major numerical languages for scientists. Additionally, scilab(1) [http://www.scilab.org] is a free Matlab\-like software, which integrates a high compatibility (syntax, functions, graphics, user interfaces).
Many IDL programs have been developed in large facilities studying astrophysics, geology,medicine, particle and nuclear physics.
On the other hand, Matlab is widely used in smaller laboratories (universities, national scientific agencies, ...). In order to extend the portability of such IDL programs, we have developped an automatic translator from IDL to Matlab and Scilab.
.PP 
This translator works as follow:
.TP 
  * analyze and formalize the IDL grammar
.TP 
  * use lex/yacc to parse the original IDL source code to translate
.TP 
  * generate a translated equivalent program, using Matlab grammar (and all variable types)
.TP 
  * use a hash table for library functions.
    This library is quite big, and covers the most useful IDL/Matlab functions, particularly mathematical functions, basic plotting functions, widgets/uicontrols.
.PP 
IDL2Matlab works on all systems (Unix/Linux/Windows). It is built upon lex/yacc and programmed in C.
The usage of the translator is done through the command line.
.PP 
Installation of the package is done with the usual procedure: ./configure; make; make install on Linux/Unix systems. An installation script is provided for Windows systems.
.P
.SH "OPTIONS"
.TP 
\-s
for script files translation
.TP 
\-S string
for string translation
.TP 
\-q 
to hide messages
.TP 
\-w 
to stop warnings writing
.TP 
\-t 
to print the abstract tree
.TP 
\-V 
to get idl2matlab version
.TP 
\-Tx
to get x spaces for a tabulation x<10
.TP 
\-f
to translate only one function
.TP 
\-C 
to translate for scilab(1) (default : matlab(1))
.SH "EXAMPLES"
.TP 
  idl2matlab \-qw essai.pro outDir/
.TP 
  idl2matlab \-t essai.pro outDir/ > out
.TP 
  idl2matlab \-S "print,'IDL2Matlab'"
.TP 
  idl2matlab \-T4 essai.pro outDir/ > out
.TP 
  idl2matlab \-f essai.pro functionName outDir/ > out
.SH "ENVIRONMENT"
The
.B IDL2MATLAB
environment variable may be set to define the IDL2Matlab library location. This is /usr/local/lib/idl2matlab on Linux/BSD systems, and C:\idl2matlab on Windows systems. Executables are usually located in /usr/local/bin on Linux/BSD systems, and in C:\idl2matlab\bin  on Windows systems.
.SH "BUGS"
GCC >= 4 installs a stack protector which leads to a buffer overflow error. 
Add the '\-fno\-stack\-protector' and remove the optimization '\-O2' to the CFLAGS variable in the Makefile.

idl2matlab is an experimental tool. The IDL grammar handled by idl2matlab is not fully qualified, and you should expect small translation errors. Additionally, the Library is certainly unstable and incomplete. PV\-Wave translation should work partially.
Graphic functions (plot, surf, widgets) are incomplete. Translated code is not guaranteed to work. IDL objects are not translated.
.SH "EXIT STATUS"
idl2matlab returns 1 in case of error, 0 when all went fine
.SH "AUTHOR"
project leaders: Emmanuel FARHI and Didier RICHARD.
team1: AZIZI MOURIER Karim, BENZEGHIOUA  ABDESLAM, GARDON Lucien, SYLVESTRE  Nadege.                
team2: BOURTEMBOURG Reynald, CORTINA Stephane, SZCZUCZAK  Nadege.   

Designed by the Computing for Science group
Institut Laue Langevin, BP 156
F\-38042 Grenoble cedex 9, FRANCE
.SH "SEE ALSO"
.P
Institut Laue Langevin at [http://www.ill.eu/]
.P
IDL at [http://www.ittvis.com/]
.P
Matlab at [http://www.mathworks.com/]
.P
Scilab at [http://www.scilab.org]
.P
matlab(1), idl(1), scilab(1), octave(1)


CHANGES
/*******************************************************************************
*
* IDL2Matlab and Scilab translator
*   Computing for Science group
*   Institut Laue langevin, BP 156
*   F-38042 Grenoble cedex 9, FRANCE
*
* Documentation: CHANGES
*
* %Identification
* Written by: EF, DR
* Date: 2003-2009
* Origin: ILL

        Release 0.0 (28 jan 2003) dev (alpha)
        Release 1.0 (31 jul 2003) dev (alpha)
        Release 1.1 (17 Nov 2003) dev (alpha)
        Release 1.2 (11 Mar 2004) dev (alpha)
        Release 1.3 (20 Mar 2006) dev (alpha)
        Release 1.4 (13 Jun 2007) beta
        Release 1.5 (21 jan 2009) final
        Release 1.6 (01 may 2013) fix buffer overflow
*
* This file is part of IDL2Matlab  version 1.6, released May 1st, 2013.
* It gives a 'changes' list from the beginning of the project
*
*******************************************************************************/

01 May 2013: fix buffer overflow error with gcc >= 4 (idl2matlab 1.6)

31 July 2003: first dev-alpha release (idl2matlab-31_Jul_2003.zip)
-------------------------------------
Supplies an Intel/Linux binary (idl2matlab).
Translate basic scripts (computations) but not efficient for graphics

17 Nov 2003: second dev-alpha release (idl2matlab-17_Nov_2003.zip)
-------------------------------------
Has many enhancements/bug corrections in grammar and code generation
Also is now supplied with a complete Matlab library, which may not work 100 %
A Scilab library is given as experimental...

11 march 2004: third dev-alpha release (idl2matlab-11_Mar_2004.zip)
-------------------------------------
Includes binaries for linux, OSF and SGI machines
With full libraries

20 march 2006: forth dev-alpha release (idl2matlab-20_Mar_2006.zip)
-------------------------------------
small bug fix. Better localization of library.

13 June 2007: beta release (idl2matlab-1.4)
-------------------------------------
better portability. Windows and Linux executables.

21 January 2009: final release (idl2matlab-1.5)
-------------------------------------
Improved Linux and Windows installation procedures



lib/idl2matlab.m
% idl2matlab(inputFile, outputDirectory, options, ...) translates IDL source code into Matlab
% input arguments must be of type char (strings). 
% There is no output parameter (translated code is generated as separate files).
%
% Options :
%   -s for script files translation
%   -S for string translation
%   -q to hide messages (quiet mode)
%   -w to stop warnings writing
%   -t to print the abstract tree (long output)
%   -V to get idl2Matlab version
%  -A to get idl2Matlab authors
%  -Tx to get x spaces for a tabulation x<10
%  -f to translate only one function
%  -C to translate in Scilab (default : Matlab)
%
%Examples :
%  idl2matlab -qw essai.pro outDir/
%  idl2matlab -t essai.pro outDir/ > out
%  idl2matlab -S "print,'IDL2Matlab'"
%  idl2matlab -T4 essai.pro outDir/
%  idl2matlab -f essai.pro functionName outDir/ > out
%  idl2matlab -C essai.pro outDir/

% idl2matlab version 1.5 090120

function idl2matlab(args)

if length(getenv('IDL2MATLAB'))
	addpath(genpath(getenv('IDL2MATLAB')));
end

if nargin == 0, args=''; end
if iscellstr(args)
  for index=1:length(args)
    idl2matlab(args{index});
  end
  return
end

if ispc
	idl2matlab_exe = 'idl2matlab.exe';
else     
	idl2matlab_exe = 'idl2matlab';
	addpath(genpath('/usr/local/lib/idl2matlab'));
end

% search executable 

while 1
  % try with system wide command
	exec = idl2matlab_exe;
	[s,w] = system(exec);
	if ~s, break; end
	% other locations
	exec = [ getenv('IDL2MATLAB') filesep idl2matlab_exe ];
	if isexec(exec), break; end
	exec = [ getenv('IDL2MATLAB') filesep 'bin' filesep idl2matlab_exe ];
	if isexec(exec), break; end
	exec = [ getenv('HOME') filesep 'bin' filesep idl2matlab_exe ];
	if isexec(exec), break; end
	exec = [ fileparts(which(mfilename)) filesep 'bin' filesep idl2matlab_exe ]; 
	if isexec(exec), break; end
	exec = [ fileparts(which(mfilename)) filesep idl2matlab_exe ]; 
	if isexec(exec), break; end
	exec = [ fileparts(which(mfilename)) filesep '..' filesep idl2matlab_exe ]; 
	if isexec(exec), break; end
	exec = [ fileparts(which(mfilename)) filesep '..' filesep 'bin' filesep idl2matlab_exe ]; 
	if isexec(exec), break; end
	exec = [ fileparts(which(mfilename)) filesep '..' filesep '..' filesep 'bin' filesep idl2matlab_exe ]; 
	if isexec(exec), break; end
	exec = '';
	break;
end
% execute command
if ~isempty(exec)
	exec = [ exec ' ' args ];
	disp(exec);
	[s,w] = system(exec);
else s=1; w='executable not found';
end
disp(w)

if s
	error('idl2matlab: failed to exectute IDL2Matalab');
else
	i2m_init;
end

% -----------------------------------------------------------------------------

function ret=isexec(file)
% test if single file exists
	ret = 0;
	d = dir(file);
	if length(d) ~= 1, return; end
	if d.isdir, return; end
	ret = 1;
	return

